pkg/reconciler/v1alpha1/taskrun/taskrun.go

```diff
@@ -254,60 +258,82 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
-	var pod *corev1.Pod
-	if tr.Status.PodName != "" {
-		pod, err = c.KubeClientSet.CoreV1().Pods(tr.Namespace).Get(tr.Status.PodName, metav1.GetOptions{})
-		if err != nil {
-			c.Logger.Error("Error getting pod %q: %v", tr.Status.PodName, err)
-			return err
-		}
-	} else {
-		pvc, err := c.KubeClientSet.CoreV1().PersistentVolumeClaims(tr.Namespace).Get(tr.Name, metav1.GetOptions{})
-		if errors.IsNotFound(err) {
-			// Create a persistent volume claim to hold Build logs
-			pvc, err = createPVC(c.KubeClientSet, tr)
-			if err != nil {
-				return fmt.Errorf("Failed to create persistent volume claim %s for task %q: %v", tr.Name, err, tr.Name)
-			}
-		} else if err != nil {
-			c.Logger.Errorf("Failed to reconcile taskrun: %q, failed to get pvc %q: %v", tr.Name, tr.Name, err)
-			return err
-		}
-		// Build pod is not present, create build pod.
-		pod, err = c.createBuildPod(ctx, tr, rtr.TaskSpec, rtr.TaskName, pvc.Name)
-		if err != nil {
-			// This Run has failed, so we need to mark it as failed and stop reconciling it
-			var msg string
-			if tr.Spec.TaskRef != nil {
-				msg = fmt.Sprintf("References a Task %s that doesn't exist: ", fmt.Sprintf("%s/%s", tr.Namespace, tr.Spec.TaskRef.Name))
-			} else {
-				msg = fmt.Sprintf("References a TaskSpec with missing information: ")
-			}
-			tr.Status.SetCondition(&duckv1alpha1.Condition{
+	var after *duckv1alpha1.Condition
+	before := tr.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+
+	if !tr.Status.StartTime.IsZero() && tr.Spec.PipelineRunTimeout != nil {
+		if time.Now().After(tr.Spec.PipelineRunTimeout.Time) {
+			timeout := tr.Spec.PipelineRunTimeout.Time.Sub(tr.Status.StartTime.Time)
+			runtime := time.Since(tr.Status.StartTime.Time)
+			c.Logger.Infof("TaskRun %q has timed out(runtime %s over %s)", tr.Name, runtime, timeout)
+
+			timeoutMsg := fmt.Sprintf("TaskRun %q failed to finish within %q", tr.Name, timeout.Round(time.Second))
+			after = &duckv1alpha1.Condition{
 				Type:    duckv1alpha1.ConditionSucceeded,
 				Status:  corev1.ConditionFalse,
-				Reason:  reasonCouldntGetTask,
-				Message: fmt.Sprintf("%s %v", msg, err),
-			})
-			c.Recorder.Eventf(tr, corev1.EventTypeWarning, "BuildCreationFailed", "Failed to create build pod %q: %v", tr.Name, err)
-			c.Logger.Errorf("Failed to create build pod for task %q :%v", err, tr.Name)
-			return nil
+				Reason:  reasonTimedOut,
+				Message: timeoutMsg,
+			}
+			tr.Status.SetCondition(after)
+			// TODO: Actually cancel the Build - see https://github.com/knative/build/pull/510 and https://github.com/knative/build-pipeline/issues/272
 		}
 	}
-	if err := c.tracker.Track(tr.GetBuildPodRef(), tr); err != nil {
-		c.Logger.Errorf("Failed to create tracker for build pod %q for taskrun %q: %v", tr.Name, tr.Name, err)
-		return err
-	}
 
-	before := tr.Status.GetCondition(duckv1alpha1.ConditionSucceeded)
+	// TODO: Skipping all things pod/build-related if we've timed out, but this probably should be better.
+	if after == nil {
```

Comment:
i feel like `after == nil` is a bit hard to understand, i wonder if there's some way this could be something like:

```
if hasTimedOut() {
  // do timeout stuff
} else {
  // do all this pod stuff
}
```

Reply:
or even like

```
if hasTimedOut() {
   // set some status stuff
   // return
}

do pod stuff
```